monai/data/meta_obj.py

```diff
@@ -0,0 +1,262 @@
+    won't be tracked.
+
+    If both `set_track_meta` and `set_track_transforms` are set to
+    `False`, then standard data objects will be returned (e.g., `torch.Tensor` and
+    `np.ndarray`) as opposed to our enhanced objects.
+
+    By default, this is `True`, and most users will want to leave it this way. However,
+    if you are experiencing any problems regarding transforms, and aren't interested in
+    preserving transforms, then you can disable it.
+    """
+    global _TRACK_TRANSFORMS
+    _TRACK_TRANSFORMS = val
+
+
+def get_track_meta() -> bool:
+    """
+    Return the boolean as to whether meta data is tracked. If `True`, meta data will be
+    associated its data by using subclasses of `MetaObj`. If `False`, then data will be
+    returned with empty meta data.
+
+    If both `set_track_meta` and `set_track_transforms` are set to
+    `False`, then standard data objects will be returned (e.g., `torch.Tensor` and
+    `np.ndarray`) as opposed to our enhanced objects.
+
+    By default, this is `True`, and most users will want to leave it this way. However,
+    if you are experiencing any problems regarding meta data, and aren't interested in
+    preserving meta data, then you can disable it.
+    """
+    return _TRACK_META
+
+
+def get_track_transforms() -> bool:
+    """
+    Return the boolean as to whether transforms are tracked. If `True`, applied
+    transforms will be associated its data by using subclasses of `MetaObj`. If `False`,
+    then transforms won't be tracked.
+
+    If both `set_track_meta` and `set_track_transforms` are set to
+    `False`, then standard data objects will be returned (e.g., `torch.Tensor` and
+    `np.ndarray`) as opposed to our enhanced objects.
+
+    By default, this is `True`, and most users will want to leave it this way. However,
+    if you are experiencing any problems regarding transforms, and aren't interested in
+    preserving transforms, then you can disable it.
+    """
+    return _TRACK_TRANSFORMS
+
+
+class MetaObj:
+    """
+    Abstract base class that stores data as well as any extra meta data and an affine
+    transformation matrix.
+
+    This allows for subclassing `torch.Tensor` and `np.ndarray` through multiple
+    inheritance.
+
+    Meta data is stored in the form of of a dictionary. Affine matrices are stored in
+    the form of e.g., `torch.Tensor` or `np.ndarray`.
+
+    We store the affine as its own element, so that this can be updated by
+    transforms. All other meta data that we don't plan on touching until we
+    need to save the image to file lives in `meta`.
+
+    Behavior should be the same as extended class (e.g., `torch.Tensor` or `np.ndarray`)
+    aside from the extended meta functionality.
+
+    Copying of information:
+        * For `c = a + b`, then auxiliary data (e.g., meta data) will be copied from the
+        first instance of `MetaObj`.
+    """
+
+    _meta: dict
+    _affine: Any
```

Comment:
could you make this `_affine` an item in `_meta`? I think with the `affine` property and getter, it's already very easy to use. this underlying structure should be simple. or is there a reason to make `_affine` special?

Reply:
I agree with @wyli , don't see any obvious reason to split `affine` and `meta dict`.

Thanks.

Reply:
I tried these APIs with more toy examples, it seems they don't work with the default collate...

Reply:
No they don't currently work with collate, but if you check my [full PR](https://github.com/rijobro/MONAI/blob/MetaTensorDev/monai/data/utils.py#L358-L361), I updated our default collation and decollation to support this. 

It's a shame that there's not a more general solution, but by the time the `default_collate` arrives in `__torch_function__`, it's `torch.stack` that's being called. We have no way of differentiating between `torch.stack` that's being called by itself versus when it's being called because of collation, so can't put that sort of logic inside the `MetaTensor` class itself.

Reply:
If `affine` lived inside of `meta`, then we would access it with `im.meta["affine"]`. We're back to using strings again. And if a lot of our transforms are going to update the affine, then `im.affine` is a lot more succinct and less prone to errors than `im.meta["affine"]`. I'd rather leave it as it is, but I am happy to change it if I'm outvoted.

Reply:
Also with `affine` as its own object, it's easy to ensure that the affine and the image are on the same device, which will be useful when transforms update it.

Reply:
perhaps misunderstanding here... I completely agree with the `im.affine` via the property interface, the comment is about making the internal data structure more consistent (merging `_affine` into `_meta`, not `affine` into `meta`). Essentially we can have
```py
@property
def affine(self) -> Any:
    return self._meta['affine']
```

Reply:
> No they don't currently work with collate, but if you check my [full PR](https://github.com/rijobro/MONAI/blob/MetaTensorDev/monai/data/utils.py#L358-L361), I updated our default collation and decollation to support this.
> 
> It's a shame that there's not a more general solution, but by the time the `default_collate` arrives in `__torch_function__`, it's `torch.stack` that's being called. We have no way of differentiating between `torch.stack` that's being called by itself versus when it's being called because of collation, so can't put that sort of logic inside the `MetaTensor` class itself.

interesting, do you think implementing an internal flag such as 'self._meta["is_batch"]' can bring more flexibilities in `__torch__function__`?

Reply:
So would the constructor be `MetaTensor(x, affine=foo, meta=bar)` or `MetaTensor(x, meta=bar)` (where the affine already lives inside of `meta`)?

If the former, what do we do if the key `affine` already exists inside of `meta`? If the latter, do we have to check inside of `meta` and if `affine` is missing then we set it to be the default? I'm struggling to see how either is cleaner.

Reply:
I think we use the former, when both `meta` has an affine and `affine=foo` is provided, we raise a warning? we may want to validate the `meta=bar` against the actual `x` anyway? such as spatial shape, data type...

Reply:
Ok. By the way, I propose that we remove any redundant information, such as spatial shape and data type, from meta. Both of these can be extracted from the image directly and only serve to cause confusion when the two inevitably drift out of sync.

Reply:
I see, that's great!